### PR TITLE
Added --window-image option to swaylock

### PR DESCRIPTION
**Pull request by nuew**
This allows setting window specific images.

Also Changed:
  - The colored background is _always_ drawn, to prevent non-fullscreen
    images from leaking information.
  - Added function list_arbitrary_insert, which allows insertion to a
    list at an arbitrary location less than the capacity. It silently
    fails otherwise.

### CODE REVIEW

**Comment by mikkeloscar** [comment]
Don't you mean output specific images?

> The colored background is always drawn, to prevent non-fullscreen images from leaking information.

I think this should be up to the user. He/she can just use scaling mode fill to make sure the image covers the entire screen, but some users might not want that. For instance it's also possible to use a semi transparent overlay color if so desired.

File: swaylock/main.c [hunk 1]
```diff
@@ -249,7 +283,38 @@ int main(int argc, char **argv) {
...
+		case 'w':
+			if(images.type == NOL_NONE) {
+				images = (struct none_one_list) { NOL_LIST, create_list(), create_list() };
+				for(i = 0; i <= ((list_t*) images.a)->capacity; ++i) {
+					// this isn't perfect - if you hav>e more than 10 monitors there's a small chance of
+					// a segfault from uninitialized data. I honestly don't think enough people have more than
+					// 10 monitors to care about that though.
+					((list_t*) images.a)->items[i] = NULL;
+					((list_t*) images.b)->items[i] = NULL;
+				}
```
Comment (mikkeloscar) [inline swaylock/main.c]: I think you would be better of using an array if you wan't to encode the output indices in the structure. A list is not really meant for that kind of thing, hence you end up doing all this :)

You could either create a fix sized array, limiting the number of outputs, or you could allocate it from `registry->outputs->length`.

File: common/list.c [hunk 2]
```diff
@@ -47,6 +47,18 @@ void list_insert(list_t *list, int index, void *item) {
...
+void list_arbitrary_insert(list_t *list, int index, void *item) {
+	list_resize(list);
+	if(index > list->capacity) {
+		return;
+	}
+	if(list->length < index) {
+		list->length = index;
+	}
+	list->items[index] = item;
+}
```
Comment (SirCmpwn) [inline common/list.c]: I don't understand why this is necessary over list_insert.

**Comment by SirCmpwn** [comment]
I would prefer to associate this with outputs rather than windows, and I think the implementation could be better.

**Comment by nuew** [comment]
I think I'm going to fully rewrite this with your advice in mind.

The reason I added list_arbitrary_insert was because list_insert `memset`s the array locations after it.

**Comment by SirCmpwn** [comment]
It does a memmove, which is necessary to, you know, insert things. Your new function wouldn't insert anything, it'd add an invalid pointer to the end of the array and then overwrite something that's already there.

**Comment by nuew** [comment]
Huh... worked for me. Maybe it was coincidence. Doesn't matter anyways, I'm not using the list anymore.

**Comment by nuew** [comment]
I can't figure out how to find the index of the output from it's name. I don't think it's actually available to clients (everything else is just using IPC, and I guess I could implement a method to send the name and get back the index, but...). Anyways, I guess I'll close this.

**Comment by SirCmpwn** [comment]
You can just do an IPC_GET_OUTPUTS and they'll be ordered according to the index.

**Comment by nuew** [comment]
Wow... I guess I'm an idiot.

**Comment by nuew** [comment]
OK, I've completed a fixed version. I've pushed it to nuew:master, but because it's almost completely rewritten, I'd recommend pulling https://github.com/nuew/sway/tree/display-images instead.

**Comment by SirCmpwn** [comment]
Can you make a new pull request?

**Pull request closed without merging**